### DISCUSSION TITLE
How can I emit to a specific room in disconnect handler?

### DISCUSSION THREAD

**Discussion by zc97** [miguelgrinberg/Flask-SocketIO#1743] [Q&A] [answered]
I try to detect the disconnection of the clients and inform it to all the clients in the room. The server should handle the disconnection event and emit an announcement event back to all other clients that are in the same room.  Then I write a disconnect hander on the server-side:

```@socketio.on('disconnect')
def handle_disconnect():
    # emit to the room is not working
    # emit('disconnect_announcement', session.get('username'), to=session.get('roomid'))
    emit('disconnect_announcement', session.get('username'), broadcast=True)  # broadcast is working
    print('connection of {} in room {} is lost'.format(session.get('username'), session.get('roomid')))
```

On the client-side, I write a handler for the 'disconnect_announcement' event and add the disconnection information to $messages:

```
socket.on('disconnect_announcement', function(data) {
            console.log('Received disconnect_announcement');
            $("#messages").append('<li>'+ data +' has lost the connection</li>');
        });
```

As a result, if I emit the 'disconnect_announcement' event with `to = session.get('roomid')`, other clients did not handle the event, if I changed the emit method to broadcast,  clients will handle the event. However, the server broadcasts it to clients that are in every room, which is not expected.

**Comment by miguelgrinberg** [comment, marked as answer]
Is the value of `session.get('roomid')` correct? Can you show me what your terminal shows for the `print` statement you have in the disconnect handler?

**Reply by zc97** [reply]
Hi, @miguelgrinberg, you are right the problem is the type of room id, it should be the <class 'str'> in python. Thank you for the answer!
